### PR TITLE
Implement ballerina toml parser

### PR DESCRIPTION
**Pull request by NatashaWso2**

### CODE REVIEW

File: misc/toml-parser/pom.xml [hunk 1]
```diff
@@ -0,0 +1,59 @@
+<!--
+  ~ Copyright (c) 2017, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (anupama-pathirage) [inline misc/toml-parser/pom.xml]: Incorrect year?

File: misc/toml-parser/src/main/java/org/ballerinalang/toml/model/Central.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+ *  KIND, either express or implied.  See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ */
+package org.ballerinalang.toml.model;
+
+/**
+ * Defines the properties used in central.
+ *
+ * @since 0.964
```
Comment (anupama-pathirage) [inline misc/toml-parser/src/main/java/org/ballerinalang/toml/model/Central.java]: Incorrect version?

File: misc/toml-parser/src/main/java/org/ballerinalang/toml/parser/ManifestBuildListener.java [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+ * @since 0.964
+ */
+public class ManifestBuildListener extends TomlBaseListener {
+    private final Manifest manifest;
+    private final SingletonStack<String> currentKey = new SingletonStack<>();
+    private Dependency dependency;
+    private String currentHeader = null;
+
+    /**
+     * Cosntructor with the manifest object.
```
Comment (anupama-pathirage) [inline misc/toml-parser/src/main/java/org/ballerinalang/toml/parser/ManifestBuildListener.java]: Typo?

File: misc/toml-parser/src/main/java/org/ballerinalang/toml/parser/ManifestBuildListener.java [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+     */
+    @Override
+    public void exitInlineTable(TomlParser.InlineTableContext ctx) {
+        setDependancyAndPatches();
+    }
+
+    /**
+     * Add the dependencies and patches to the manifest object.
+     */
+    private void setDependancyAndPatches() {
```
Comment (anupama-pathirage) [inline misc/toml-parser/src/main/java/org/ballerinalang/toml/parser/ManifestBuildListener.java]: Typo? Need to be setDependencyAndPatches

File: misc/toml-parser/src/main/java/org/ballerinalang/toml/util/SingletonStack.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+ * @param <T>
+ * @since 0.964
+ */
+public class SingletonStack<T> {
+    private T content;
+
+    /**
+     * Push object to the content.
+     *
+     * @param obj
```
Comment (anupama-pathirage) [inline misc/toml-parser/src/main/java/org/ballerinalang/toml/util/SingletonStack.java]: Shall we add a description  here?

File: misc/toml-parser/src/main/resources/Toml.g4 [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+
+newline : '\n' | '\r\n';
+
+// Comment
+
+COMMENT : '#' .*? ('\n' | '\r' | '\t') -> channel(2) ;
+
+//Key-Value pairs
+
+keyval : key keyvalSep val;
```
Comment (anupama-pathirage) [inline misc/toml-parser/src/main/resources/Toml.g4]: Shall we rename this to keyVal or keyValue. same goes with keyvalSep - keyValSep or KeyValueSep

File: misc/toml-parser/src/test/java/org/ballerinalang/toml/parser/ManifestProcessorTest.java [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+
+import org.ballerinalang.toml.model.Manifest;
+import org.ballerinalang.toml.parser.ManifestProcessor;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+import java.io.IOException;
+
+/**
+ * Test class to populate Manifest object by reading the toml
```
Comment (anupama-pathirage) [inline misc/toml-parser/src/test/java/org/ballerinalang/toml/parser/ManifestProcessorTest.java]: End with a period.

File: misc/toml-parser/src/test/java/org/ballerinalang/toml/parser/ManifestProcessorTest.java [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+    }
+
+    @Test(description = "Description in package section has an effect")
+    public void testLicenseDescription() throws IOException {
+        Manifest manifest = ManifestProcessor.parseTomlContentFromString("[package] \n " +
+                "license = \"MIT OR Apache-2.0\"");
+        Assert.assertEquals(manifest.getLicense(), "\"MIT OR Apache-2.0\"");
+    }
+
+    @Test(description = "One dependancy added to the dependancies section has an effect")
```
Comment (anupama-pathirage) [inline misc/toml-parser/src/test/java/org/ballerinalang/toml/parser/ManifestProcessorTest.java]: Typo - should be dependency

File: misc/toml-parser/src/test/java/org/ballerinalang/toml/parser/ManifestProcessorTest.java [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+    public void testSingleDependancies() throws IOException {
+        Manifest manifest = ManifestProcessor.parseTomlContentFromString("[dependencies] \n " +
+                "string-utils = {location = \"src/string-utils\", version = \"0.15\"} \n");
+        Assert.assertEquals(manifest.getDependencies().get(0).getPackageName(), "string-utils");
+        Assert.assertEquals(manifest.getDependencies().get(0).getVersion(), "\"0.15\"");
+        Assert.assertEquals(manifest.getDependencies().get(0).getLocation(), "\"src/string-utils\"");
+    }
+
+    @Test(description = "Empty dependancy added to the dependancies section has no effect")
```
Comment (anupama-pathirage) [inline misc/toml-parser/src/test/java/org/ballerinalang/toml/parser/ManifestProcessorTest.java]: Typo - should be dependency. check the whole file

File: misc/toml-parser/src/test/resources/testng.xml [hunk 10]
```diff
@@ -0,0 +1,28 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+Copyright (c) 2016, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (anupama-pathirage) [inline misc/toml-parser/src/test/resources/testng.xml]: Incorrect version?

File: misc/toml-parser/.gitignore [hunk 11]
```diff
@@ -0,0 +1,21 @@
+*.class
```
Comment (pubudu91) [inline misc/toml-parser/.gitignore]: Do we need a separate .gitignore file for this module?

File: misc/toml-parser/pom.xml [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+    <modelVersion>4.0.0</modelVersion>
+    <artifactId>toml-parser</artifactId>
+    <packaging>jar</packaging>
+
+    <name>Ballerina - Toml Parser</name>
+
+    <dependencies>
+        <dependency>
+            <groupId>org.ballerinalang</groupId>
+            <artifactId>ballerina-core</artifactId>
```
Comment (pubudu91) [inline misc/toml-parser/pom.xml]: Use Antlr dependency directly.

File: misc/toml-parser/pom.xml (hunk 1 above)
Comment (NatashaWso2) [reply misc/toml-parser/pom.xml]: Fixed with 82dd533

File: misc/toml-parser/src/main/java/org/ballerinalang/toml/model/Central.java (hunk 2 above)
Comment (NatashaWso2) [reply misc/toml-parser/src/main/java/org/ballerinalang/toml/model/Central.java]: Removed all the custom listeners from the toml-parser

File: misc/toml-parser/src/main/java/org/ballerinalang/toml/parser/ManifestBuildListener.java (hunk 3 above)
Comment (NatashaWso2) [reply misc/toml-parser/src/main/java/org/ballerinalang/toml/parser/ManifestBuildListener.java]: Removed all the custom listeners from the toml-parser

File: misc/toml-parser/src/main/java/org/ballerinalang/toml/parser/ManifestBuildListener.java (hunk 4 above)
Comment (NatashaWso2) [reply misc/toml-parser/src/main/java/org/ballerinalang/toml/parser/ManifestBuildListener.java]: Removed all the custom listeners from the toml-parser

File: misc/toml-parser/src/main/java/org/ballerinalang/toml/util/SingletonStack.java (hunk 5 above)
Comment (NatashaWso2) [reply misc/toml-parser/src/main/java/org/ballerinalang/toml/util/SingletonStack.java]: Removed all the custom listeners from the toml-parser

File: misc/toml-parser/src/main/resources/Toml.g4 (hunk 6 above)
Comment (NatashaWso2) [reply misc/toml-parser/src/main/resources/Toml.g4]: Fixed with 46e540d

File: misc/toml-parser/src/test/java/org/ballerinalang/toml/parser/ManifestProcessorTest.java (hunk 7 above)
Comment (NatashaWso2) [reply misc/toml-parser/src/test/java/org/ballerinalang/toml/parser/ManifestProcessorTest.java]: Removed all the custom listeners from the toml-parser

File: misc/toml-parser/src/test/java/org/ballerinalang/toml/parser/ManifestProcessorTest.java (hunk 8 above)
Comment (NatashaWso2) [reply misc/toml-parser/src/test/java/org/ballerinalang/toml/parser/ManifestProcessorTest.java]: Removed all the custom listeners from the toml-parser

File: misc/toml-parser/src/test/java/org/ballerinalang/toml/parser/ManifestProcessorTest.java (hunk 9 above)
Comment (NatashaWso2) [reply misc/toml-parser/src/test/java/org/ballerinalang/toml/parser/ManifestProcessorTest.java]: Removed all the custom listeners from the toml-parser

File: misc/toml-parser/src/test/resources/testng.xml (hunk 10 above)
Comment (NatashaWso2) [reply misc/toml-parser/src/test/resources/testng.xml]: Removed file from the toml-parser

File: misc/toml-parser/findbugs-exclude.xml [hunk 13]
```diff
@@ -0,0 +1,25 @@
+<!--
+  ~ Copyright (c) 2016, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (anupama-pathirage) [inline misc/toml-parser/findbugs-exclude.xml]: Incorrect year?

**Pull request merged**